python/mxnet/io.py

```diff
@@ -19,6 +19,51 @@
 from .ndarray import concatenate
 
 
+class LayoutMapper(object):
+    """A helper class to decide data layouts (which axis is batch_size dimension)."""
+    def get_data_layout(self, name):
+        """Get data layout.
+
+        Parameters
+        ----------
+        name : str
+            The name of the blob (could be data, label or output names).
+
+        Returns
+        -------
+        An axis indicating the batch_size dimension. When data-parallelism is
+        used, the data will be automatically split and concatenate along the batch_size
+        dimension. Axis can be -1, which means the array thing will be copied for each
+        data-parallelism device.
+        """
+        raise NotImplementedError()
+
+
+class DefaultLayoutMapper(LayoutMapper):
```

Comment:
Is this necessary?
You can determine this when computing slices.
instead of one dimensional slices, use multi dimensional slice like [(a,b,c),(a,b,c), ...]
Then change copy slice from and assign slice to to slice and `__setitem__`
Feels like a much cleaner way of doing it